src/fread.c

```diff
@@ -304,7 +347,8 @@ static inline bool end_of_field(const char *ch)
   // default, and therefore characters in the range 0x80-0xFF are negative.
   // We use eol() because that looks at eol_one_r inside it w.r.t. \r
   // \0 (maybe more than one) before eof are part of field and do not end it; eol() returns false for \0 but the ch==eof will return true for the \0 at eof.
-  return *ch == sep || ((uint8_t)*ch <= 13 && (ch == eof || eol(&ch)));
+  // Comment characters terminate a field immediately and take precedence over separators.
+  return *ch == sep || ((uint8_t)*ch <= 13 && (ch == eof || eol(&ch))) || (commentChar && *ch == commentChar);
```

Comment:
Maybe time to split it out more, WDYT? (I don't think there is any performance implication)

```suggestion
  if (*ch == sep) return true;
  if (((uint8_t)*ch <= 13 && (ch == eof || eol(&ch))) return true;
  if (!commentChar) return false;
  return *ch == commentChar;
```